test/servername_test.c

```diff
@@ -0,0 +1,241 @@
+    PACKET pkt = {0}, pkt2 = {0}, pkt3 = {0}, pkt4 = {0}, pkt5 = {0};
+    unsigned int servname_type = 0, type = 0;
+    int ret = 0;
+
+    len = BIO_get_mem_data(bio, (char **)&data);
+    if (!TEST_true(PACKET_buf_init(&pkt, data, len))
+               /* Skip the record header */
+            || !PACKET_forward(&pkt, SSL3_RT_HEADER_LENGTH))
+        goto end;
+
+    /* Skip the handshake message header */
+    if (!TEST_true(PACKET_forward(&pkt, SSL3_HM_HEADER_LENGTH))
+               /* Skip client version and random */
+            || !TEST_true(PACKET_forward(&pkt, CLIENT_VERSION_LEN
+                                               + SSL3_RANDOM_SIZE))
+               /* Skip session id */
+            || !TEST_true(PACKET_get_length_prefixed_1(&pkt, &pkt2))
+               /* Skip ciphers */
+            || !TEST_true(PACKET_get_length_prefixed_2(&pkt, &pkt2))
+               /* Skip compression */
+            || !TEST_true(PACKET_get_length_prefixed_1(&pkt, &pkt2))
+               /* Extensions len */
+            || !TEST_true(PACKET_as_length_prefixed_2(&pkt, &pkt2)))
+        goto end;
+
+    /* Loop through all extensions for SNI */
+    while (PACKET_remaining(&pkt2)) {
+        if (!TEST_true(PACKET_get_net_2(&pkt2, &type))
+                || !TEST_true(PACKET_get_length_prefixed_2(&pkt2, &pkt3)))
+            goto end;
+        if (type == TLSEXT_TYPE_server_name) {
+            if (!TEST_true(PACKET_get_length_prefixed_2(&pkt3, &pkt4))
+                    || !TEST_uint_ne(PACKET_remaining(&pkt4), 0)
+                    || !TEST_true(PACKET_get_1(&pkt4, &servname_type))
+                    || !TEST_uint_eq(servname_type, TLSEXT_NAMETYPE_host_name)
+                    || !TEST_true(PACKET_get_length_prefixed_2(&pkt4, &pkt5))
+                    || !TEST_uint_le(PACKET_remaining(&pkt5), TLSEXT_MAXLEN_host_name)
+                    || !TEST_false(PACKET_contains_zero_byte(&pkt5))
+                    || !TEST_true(PACKET_strndup(&pkt5, sni)))
+                goto end;
+            ret = 1;
+            goto end;
+        }
+    }
+end:
+    return ret;
+}
+
+static int client_setup_sni_before_state()
+{
+    SSL_CTX *ctx;
+    SSL *con = NULL;
+    BIO *rbio;
+    BIO *wbio;
+    char *hostname = NULL;
+    int ret = 0;
+
+    /* use TLS_method to blur 'side' */
+    ctx = SSL_CTX_new(TLS_method());
+    if (!TEST_ptr(ctx))
+        goto end;
+
+    con = SSL_new(ctx);
+    if (!TEST_ptr(con))
+        goto end;
+
+    /* set SNI before 'client side' is set */
+    SSL_set_tlsext_host_name(con, host);
+
+    rbio = BIO_new(BIO_s_mem());
+    wbio = BIO_new(BIO_s_mem());
+    if (!TEST_ptr(rbio)|| !TEST_ptr(wbio)) {
+        BIO_free(rbio);
+        BIO_free(wbio);
+        goto end;
+    }
+
+    SSL_set_bio(con, rbio, wbio);
+
+    if (!TEST_int_le(SSL_connect(con), 0))
+        /* This shouldn't succeed because we don't have a server! */
+        goto end;
+    if (!TEST_true(get_sni_from_client_hello(wbio, &hostname)))
```

Comment:
Side note: when I was first talking about tests that would detect if we had denied setting the sni hostname if s->server, I was thinking about having a "normal" sslapitest that actually does connect to a server SSL, and pulling the received hostname from the server SSL.  But this approach also works, and I don't want to make you rewrite things.  It may even be illustrative to have an "example test" that does this parsing.